lib/Transforms/MemoryBanking.cpp

```diff
@@ -200,11 +200,41 @@ SmallVector<Value, 4> handleGetGlobalOp(memref::GetGlobalOp getGlobalOp,
   return banks;
 }
 
+std::optional<int> getBankingDimension(std::optional<int> bankingDimensionOpt,
+                                       int64_t rank, ArrayRef<int64_t> shape) {
+  // If the banking dimension is already specified, return it.
+  // Note, the banking dimension will always be nonempty because TableGen will
+  // assign it with a default value -1 if it's not specified by the user. Thus,
+  // -1 is the sentinel value to indicate the default behavior, which is the
+  // innermost dimension with shape greater than 1.
+  if (bankingDimensionOpt.has_value() && *bankingDimensionOpt >= 0) {
+    return bankingDimensionOpt;
+  }
+
+  // Otherwise, find the innermost dimension with size > 1.
+  // For example, [[1], [2], [3], [4]] with `bankingFactor`=2 will be banked to
+  // [[1], [3]] and [[2], [4]].
+  for (int dim = rank - 1; dim >= 0; --dim) {
+    if (shape[dim] > 1) {
+      return dim;
+    }
+  }
+
+  return std::nullopt;
+}
+
 SmallVector<Value, 4> createBanks(Value originalMem, uint64_t bankingFactor,
-                                  unsigned bankingDimension) {
+                                  std::optional<int> bankingDimensionOpt) {
   MemRefType originalMemRefType = cast<MemRefType>(originalMem.getType());
+  unsigned rank = originalMemRefType.getRank();
+  ArrayRef<int64_t> shape = originalMemRefType.getShape();
+
+  auto bankingDimension = getBankingDimension(bankingDimensionOpt, rank, shape);
+  assert(bankingDimension.has_value() && bankingDimension >= 0 &&
+         "No eligible dimension for banking");
```

Comment:
Suggestion: Instead of repeating this 3 times, should we just place it directly in the `getBankingDimension` function?

Reply:
indeed, thanks!